aten/src/ATen/native/Loss.cpp

```diff
@@ -189,9 +189,12 @@ Tensor margin_ranking_loss(const Tensor& input1, const Tensor& input2, const Ten
 
 Tensor kl_div(const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
   TORCH_CHECK(!input.is_complex() && !target.is_complex(),
-              "kl_div: Complex inputs not supported.")
+              "kl_div: Complex inputs not supported.");
+  TORCH_CHECK(!at::isIntegralType(input.scalar_type(), /*include_bool*/true) &&
+              !at::isIntegralType(target.scalar_type(), /*include_bool*/true),
+              "kl_div: Integral inputs not supported.");
   auto output = log_target ? at::exp(target) * (target - input)
-                           : target * (at::log(target) - input);
+                           : at::xlogy(target, target) - target * input;
```

Comment:
Can you give more details on how you measure the regression. From local check, it shouldn't be that bad:
```python
In [6]: a = torch.rand(100)

In [7]: %timeit tmp = a * (a.log() - b)
2.59 µs ± 8.23 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [8]: %timeit tmp = torch.xlogy(a, a) - a * b
2.86 µs ± 10.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Reply:
For small inputs, the kernel is memory bound, so that's why the results are similar. See the benchmarks:
```
[ kl_div cpu (100,) ]
               |  PR | master
32 threads: -------- | ----
      fwd      |  15 |  19
      bwd      |  61 |  40
      fwd+bwd  |  82 |  71

Times are in microseconds (us).

[ kl_div cpu (1000, 1000) ]
               |   PR | master
32 threads: --------- | ----
      fwd      |  200 | 169
      bwd      |  400 | 115
      fwd+bwd  |  600 | 280

Times are in microseconds (us).

[ kl_div cuda (100,) ]
               |   PR | master
32 threads: --------- | ----
      fwd      |   35 |  34
      bwd      |  133 | 100
      fwd+bwd  |  216 | 214

Times are in microseconds (us).

[ kl_div cuda (1000, 1000) ]
               |   PR | master
32 threads: --------- | ----
      fwd      |  132 | 132
      bwd      |  485 | 261
      fwd+bwd  |  619 | 392

Times are in microseconds (us).
```
You see that in CUDA it's still memory bound. The only difference comes from calling 1 more kernel in the backward. Well worth doing this to get correctness IMO.

Here's the script:
```python
from itertools import product

import torch
from torch.utils.benchmark import Compare, Timer


def get_timer(size, device, fn, name):
    # a is in the log-space, i.e. [-inf, 0]
    # b is a probability measure, i.e. [0, 1]
    a = torch.randn(*size, device=device)
    a = -torch.nn.functional.softplus(a)
    a.requires_grad_()

    b = torch.randn(*size, device=device).sigmoid()
    b.requires_grad_()

    z = torch.nn.functional.kl_div(a, b, reduction="batchmean")
    print(name, a.shape, device)

    timer = Timer(
        fn,
        globals={"a": a, "b": b, "z": z},
        label=f"kl_div {device} {size}",
        sub_label=name,
        description="PR",
        num_threads=torch.get_num_threads())

    return timer.blocked_autorange(min_run_time=5)


def get_params():
    fns = [("fwd", "torch.nn.functional.kl_div(a, b, reduction='batchmean')"),
           ("bwd", "z.backward(retain_graph=True)"),
           ("fwd+bwd", "torch.nn.functional.kl_div(a, b, reduction='batchmean').backward()")]
    for device, (name, fn) in product(("cpu", "cuda"), fns):
        yield (100,), device, fn, name
        yield (1000, 1000), device, fn, name


compare = Compare([get_timer(*params) for params in get_params()])
compare.trim_significant_figures()
compare.print()
```

Reply:
fwiw, I joined the outputs from master and the PR by hand because I'm lazy.